agent/consul/gateway_locator.go

```diff
@@ -0,0 +1,284 @@
+		idx, err := g.runOnce(lastFetchIndex)
+		if err != nil {
+			return err
+		}
+
+		lastFetchIndex = idx
+
+		return nil
+	}, func(err error) {
+		g.logger.Error("error tracking primary and local mesh gateways", "error", err)
+	})
+}
+
+func (g *GatewayLocator) runOnce(lastFetchIndex uint64) (uint64, error) {
+	// NOTE: we can't do RPC here because we won't have a token so we'll just
+	// mostly assume that our FSM is caught up enough to answer locally.  If
+	// this has drifted it's no different than a cache that drifts or an
+	// inconsistent read.
+	queryOpts := &structs.QueryOptions{
+		MinQueryIndex:     lastFetchIndex,
+		RequireConsistent: false,
+	}
+
+	var (
+		results   []*structs.FederationState
+		queryMeta structs.QueryMeta
+	)
+	err := g.srv.blockingQuery(
+		queryOpts,
+		&queryMeta,
+		func(ws memdb.WatchSet, state *state.Store) error {
+			// Get the existing stored version of this config that has replicated down.
+			// We could phone home to get this but that would incur extra WAN traffic
+			// when we already have enough information locally to figure it out
+			// (assuming that our replicator is still functioning).
+			idx, all, err := state.FederationStateList(ws)
+			if err != nil {
+				return err
+			}
+
+			queryMeta.Index = idx
+			results = all
+
+			return nil
+		})
+	if err != nil {
+		return 0, err
+	}
+
+	g.updateFromState(results)
+
+	return queryMeta.Index, nil
+}
+
+func (g *GatewayLocator) updateFromState(results []*structs.FederationState) {
+	var (
+		local   structs.CheckServiceNodes
+		primary structs.CheckServiceNodes
+	)
+	for _, config := range results {
+		retained := retainGateways(config.MeshGateways)
+		if config.Datacenter == g.datacenter {
+			local = retained
+		}
+		// NOT else-if because conditionals are not mutually exclusive
+		if config.Datacenter == g.primaryDatacenter {
+			primary = retained
+		}
+	}
+
+	primaryAddrs := renderGatewayAddrs(primary, true)
+	localAddrs := renderGatewayAddrs(local, false)
+
+	g.gatewaysLock.Lock()
+	defer g.gatewaysLock.Unlock()
+
+	changed := false
+	primaryReady := false
+	if !lib.StringSliceEqual(g.primaryGateways, primaryAddrs) {
+		g.primaryGateways = primaryAddrs
+		primaryReady = len(g.primaryGateways) > 0
+		changed = true
+	}
+	if !lib.StringSliceEqual(g.localGateways, localAddrs) {
+		g.localGateways = localAddrs
+		changed = true
+	}
+
+	if changed {
+		g.logger.Debug(
+			"new cached locations of mesh gateways",
+			"primary", primaryAddrs,
+			"local", localAddrs,
+		)
+	}
+
+	if primaryReady {
+		select {
+		case _, open := <-g.primaryGatewaysReadyCh:
+			if open {
+				close(g.primaryGatewaysReadyCh)
+			}
+		default:
+		}
+	}
+}
+
+func retainGateways(full structs.CheckServiceNodes) structs.CheckServiceNodes {
+	out := make([]structs.CheckServiceNode, 0, len(full))
+	for _, csn := range full {
+		if csn.Service.Meta[structs.MetaWANFederationKey] != "1" {
+			continue
+		}
+
+		// only keep healthy ones
```

Comment:
🤔 do you think we need a "panic threshold" where if there are multiple gateways and ~all~ more than X% are unhealthy we panic and just route to them anyway on the logic that it's better to try than fail completely?